### PR TITLE
add env check for installation

### PR DESCRIPTION
**Pull request by raphaelgavache**
## Changes overview
- support of DD_APM_INSTRUMENTATION_ENABLED
- conditioned docker configs on docker being instrumented (need to rethink as that's extra complexity for nothing)

## Tests

| DD_APM_INSTRUMENTATION_ENABLED         | docker instrumented | host instrumented | succeed installation |
|----------------------------------------|---------------------|-------------------|----------------------|
| host                                   | ❌                  | ✅                | ✅                   |
| docker                                 | ✅                  | ❌                | ✅                   |
| all                                    | ✅                  | ✅                | ✅                   |
| not set                                | ✅                  | ✅                | ✅                   |

## Follow up tests (merging to avoid conflict and then completing)

| DD_APM_INSTRUMENTATION_ENABLED         | docker instrumented | host instrumented | succeed installation |
|----------------------------------------|---------------------|-------------------|----------------------|
| docker (docker not installed)          | ❌                  | ❌                | ❌                   |
| all (docker not installed)             | ❌                  | ✅                | ✅                   |
| unknown                                | ❌                  | ❌                | ❌                   |
| not set (docker not installed)         | ❌                  | ✅                | ✅                   |

### CODE REVIEW

File: pkg/fleet/installer/service/apm_inject.go [hunk 1]
```diff
@@ -212,3 +216,31 @@ func (a *apmInjectorInstaller) verifySharedLib(ctx context.Context, libPath stri
...
+		return false
+	default:
+		log.Warnf("Unknown value for DD_APM_INSTRUMENTATION_ENABLED: %s. Supported values are all/docker/host", execEnvs.InstallScript.APMInstrumentationEnabled)
+		return false
+	}
+}
+
+func mustInstrumentDocker(execEnvs *env.Env) bool {
+	return execEnvs.InstallScript.APMInstrumentationEnabled == env.APMInstrumentationEnabledDocker
+}
```
Comment (BaptisteFoy) [inline pkg/fleet/installer/service/apm_inject.go:246]: **🥜 Nitpick:** ‏Comments would be nice as we won't have the nice markdown table in the code

Comment (BaptisteFoy) [reply pkg/fleet/installer/service/apm_inject.go:246]: **🥜 Nitpick:** ‏Why not just pass the `APMInstrumentationEnabled` env var?

File: pkg/fleet/installer/service/apm_inject.go
Comment (BaptisteFoy) [inline pkg/fleet/installer/service/apm_inject.go:1]: We're missing the "unknown = failure" mode

File: pkg/fleet/installer/service/apm_inject.go [hunk 2]
```diff
@@ -80,61 +81,64 @@ func (a *apmInjectorInstaller) Setup(ctx context.Context) (err error) {
...
+		if err != nil {
+			return err
 		}
-	}()
+		defer func() {
+			if retErr != nil && rollbackLDPreload != nil {
+				if err := rollbackLDPreload(); err != nil {
+					log.Warnf("Failed to rollback agent config: %v", err)
+				}
+			}
```
Comment (BaptisteFoy) [inline pkg/fleet/installer/service/apm_inject.go:95]: **💭 Thought:** ‏I think it's less footgun-y for telemetry to keep `err` as the main error and user another name for rollback errors.

File: pkg/fleet/installer/service/apm_sockets.go [hunk 3]
```diff
@@ -74,19 +74,21 @@ func getSocketsPath() (string, string, error) {
...
+		return err
+	}
+	defer func() {
+		if retErr != nil && rollback != nil {
 			rollbackErr := rollback()
-			if rollbackErr != nil {
+			if err := rollback(); err != nil {
 				log.Warnf("Failed to rollback environment file: %v", rollbackErr)
 			}
 		}
```
Comment (BaptisteFoy) [inline pkg/fleet/installer/service/apm_sockets.go:90]: Same comment for `retErr`

File: pkg/fleet/internal/exec/installer_exec.go [hunk 4]
```diff
@@ -42,15 +42,13 @@ type installerCmd struct {
 }
 
 func (i *InstallerExec) newInstallerCmd(ctx context.Context, command string, args ...string) *installerCmd {
-	env := i.env.ToEnv()
```
Comment (BaptisteFoy) [inline pkg/fleet/internal/exec/installer_exec.go:45]: **🔨 Warning:** ‏You can't really remove this, as technically `i.env` can be [populated from a config file](https://github.com/DataDog/datadog-agent/blob/db5c3f8609cd02e7327b96c35f4ef5797911dd9c/pkg/fleet/env/env.go#L61-L69) (when running in the daemon for instance).

File: pkg/fleet/installer/service/apm_inject.go (hunk 2 above)
Comment (raphaelgavache) [reply pkg/fleet/installer/service/apm_inject.go:95]: you want to trigger a rollback on the return error when the defer func is called. To ensure that err does not get overiden/ scoped to a { }, the best way is to use a different err for return

File: pkg/fleet/installer/service/apm_inject.go (hunk 1 above)
Comment (raphaelgavache) [reply pkg/fleet/installer/service/apm_inject.go:246]: I'm part of the code is comment club, the function name shouldInstrument/mustInstrument are not self-explanatory?

File: pkg/fleet/internal/exec/installer_exec.go (hunk 4 above)
Comment (raphaelgavache) [reply pkg/fleet/internal/exec/installer_exec.go]: nice catch, fixed

File: pkg/fleet/installer/service/apm_inject.go (hunk 1 above)
Comment (BaptisteFoy) [reply pkg/fleet/installer/service/apm_inject.go:246]: When not taken in context, not really. Having the different cases summarised somewhere in the code would help those not familiar with the injection.

I won't block the PR on this.

**Review by BaptisteFoy — APPROVED** [review]
ok with the follow ups

**Pull request merged**